modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +572,199 @@ export class OlmAdapter extends Listenable {
+
+                if (sasCommitment) {
+                    const olmUtil = new Olm.Utility();
+                    const commitment = olmUtil.sha256(theirKey + startContent);
+
+                    olmUtil.free();
+
+                    if (sasCommitment !== commitment) {
+                        this._sendError(participant, 'OlmAdapter commitments mismatched');
+                        olmData.sasVerification.free();
+
+                        return;
+                    }
+                }
+
+                sas.set_their_key(theirKey);
+
+                const pubKey = sas.get_pubkey();
+
+                const myInfo = `${this.myId}|${pubKey}`;
+                const theirInfo = `${pId}|${theirKey}`;
+
+                const info = isInitializer ? `${myInfo}|${theirInfo}` : `${theirInfo}|${myInfo}`;
+
+                const sasBytes = sas.generate_bytes(info, OLM_SAS_NUM_BYTES);
+                const generatedSas = generateSas(sasBytes);
+
+                this.eventEmitter.emit(OlmAdapterEvents.PARTICIPANT_SAS_READY, pId, generatedSas);
+
+                if (keySent) {
+                    return;
+                }
+
+                const ack = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pubKey,
+                            uuid
+                        }
+                    }
+                };
+
+                this._sendMessage(ack, pId);
+
+                olmData.sasVerification.keySent = true;
+            } else {
+                logger.debug(`Received sas key message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-key');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_MAC: {
+            if (olmData.session) {
+                const { keys, mac, uuid } = msg.data;
+
+                if (!mac || !keys) {
+                    logger.warn('Invalid SAS MAC message');
+
+                    return;
+                }
+
+                const { sas, ed25519 } = olmData.sasVerification;
+
+                // Verify the received MACs.
+                const baseInfo = `${OLM_KEY_VERIFICATION_MAC_INFO}${pId}${this.myId}${uuid}`;
+                const keysMac = sas.calculate_mac(
+                    Object.keys(mac).sort().join(','), // eslint-disable-line newline-per-chained-call
+                    baseInfo + OLM_KEY_VERIFICATION_MAC_KEY_IDS
+                );
+
+                if (keysMac !== keys) {
+                    logger.error('SAS verification error: keys MAC mismatch');
+                    this.eventEmitter.emit(
+                        OlmAdapterEvents.PARTICIPANT_VERIFICATION_COMPLETED,
+                        pId,
+                        false,
+                        'keys MAC mismatch');
```

Comment:
let's create constants for this please, so the app can then render nice(r) messages.